### PR TITLE
Switch posticle to use an AST and make it portable

### PR DESCRIPTION
**Pull request by measlytwerp**
- [x] Remove validation (it will be moved to rustodon)
- [x] Rewrite the parser definition to capture everything (and use more built-ins)
- [x] Define and populate an AST on parse
- [x] Add `EntityTransformer` trait and a default implementation
- [x] Add a method for rendering sanitised HTML
- [x] Implement an `EntityTransformer` for rustodon that does validation

See #117.

### CODE REVIEW

**Comment by measlytwerp** [comment]
Mostly happy with how this has gone, although I feel bad for undoing some of the work you did making everything work with a simple `map` method call, I'm still new enough to rust that I'm not sure if I can do something similar with the new structure.

In the end passing around `Option<Entity>` was a little limiting, so now the `EntityTransformer` trait and `parse` methods return a `Vec<Entity>` instead. This allows you to do more complex transformations, it's not strictly necessary for `rustodon` so perhaps we can walk that bit back.

Also, it now also supports `:emoticons:`.

**Comment by measlytwerp** [comment]
Based on todays work, here's how you turn Link tokens into HTML:

```rust
impl TokenTransformer for HtmlTransformer {
    fn transform(&self, token: Token) -> Vec<Token> {
        match token {
            Token::Link(link) => vec![Token::Element(Element(
                "a".to_string(),
                Some(vec![("href".to_string(), link.0.clone())]),
                Some(link.0),
            ))],
            _ => vec![token],
        }
    }
}
```

Still need to add tests for rendering tokens, and to sanitize the output.

**Comment by measlytwerp** [comment]
HTML sanitisation is done, it was fairly simple, but I found the need to change the `TokenTransformer` trait into something more general:

```rust
use ammonia::Builder as Ammonia;

pub trait PosticleConfig {
    fn html_sanitizer(&self) -> Ammonia;
    fn transform_token(&self, token: Token) -> Vec<Token>;
}
```

There is of course a default implementation of this using `Posticle::new()` but using an implementation of the trait is as easy as `Posticle::from(CustomConfig)`. This was my first time using a `Box` amazingly.

Here's an example implementation that just turns `Token::Link`s into HTML:

```rust
impl PosticleConfig for SimpleHtml {
    fn html_sanitizer(&self) -> Ammonia {
        let mut sanitizer = Ammonia::default();

        sanitizer.tags(hashset!["a", "br"]);

        sanitizer
    }

    fn transform_token(&self, token: Token) -> Vec<Token> {
        match token {
            Token::Link(link) => vec![Token::Element(Element(
                "a".to_string(),
                Some(vec![("href".to_string(), link.0.clone())]),
                Some(link.0),
            ))],
            _ => vec![token],
        }
    }
}
```

@barzamin would be good to get your thoughts on this structure before I start writing countless tests for it, while I've been following rust for years I've not got a lot of experience and don't know all the best practices.

**Comment by measlytwerp** [comment]
I'm moving this away from the weird pattern it's using and to a builder pattern.

I'm also looking again at the `TokenTransformer` trait, I wanted to be able to use it to gather information about the content, but that means everything needs to be mutable. I'm not yet sure what it'll look like when done.

**Comment by measlytwerp** [comment]
I've settled on a reader/writer pattern as that allows the same functionality as the `TokenTransformer` but doesn't require that `Posticle` accept an instance of a trait.

**Comment by measlytwerp** [comment]
Right, I'm much happier with the shape things are in now, here's how you transform results instead of using `TokenTransformer`:

```rust
let transformer = |token| match token {
    Token::Link(link) => {
        let url = Url::parse(&link.0).unwrap();

        Token::Element(Element(
            "a".to_string(),
            Some(vec![("href".to_string(), link.0.clone())]),
            Some(url.domain().unwrap().to_string()),
        ))
    },
    _ => token,
};

let mut html_sanitizer = Builder::default();

html_sanitizer.tags(hashset!["br", "a"]);

let tokens = Reader::from("Mastodon is great! https://joinmastodon.org/")
    .map(transformer)
    .collect::<Vec<Token>>();

let html = Writer::from(tokens)
    .with_html_sanitizer(html_sanitizer)
    .to_string();
```

**Comment by barzamin** [comment]
okay, so we have an AST but it's still a flat stream of tokens, basically? hmmmm. how does this interplay with, eg, having a string like `look :chicken: at :chicken: my :chicken: cluckers :chicken:`, where we're going to want to write out a single `<p>` looking something like
```html
<p>look &#x1f414; at &#x1f414; my &#x1f414; cluckers &#x1f414;</p>
```

File: lib/posticle/src/tokens.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+impl Text {
+    pub fn render(&self, output: &mut String) {
+        output.push_str(&html_escape(&self.0));
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+/// Trusted HTML element with content that will have its entities encoded on render.
+pub struct Element(
```
Comment (barzamin) [inline lib/posticle/src/tokens.rs]: Consider naming these struct fields?

**Comment by measlytwerp** [comment]
> from what i've seen, this looks awesome! i'm going to try and review it now. not sure how far i'll get in given that i'm incredibly low-spoons atm, but i'll try!

@barzamin hope you're ok! Thanks for taking the time, as it's revealed a problem...

> okay, so we have an AST but it's still a flat stream of tokens, basically? hmmmm. how does this interplay with, eg, having a string like look :chicken: at :chicken: my :chicken: cluckers :chicken:, where we're going to want to write out a single <p> looking something like

At the minute no paragraph elements are created, instead it's outputting br elements where the line breaks are, and indeed, the AST is currently flat. I was planning to make it into an actual tree as part of the markdown work (#117), however that problem I mentioned...

By replacing the `TokenTransformer` trait with a reader/writer pattern I've lost the ability to alter tokens below the first level of the tree easily. I still think that the reader/writer pattern is the right approach for this, however I think we are going to need additional abilities on the reader to support transformations *as the tokens are generated.*

I'm going to have a look around for some prior art in this area before I commit to any course of action, but I suspect I'm going to be passing a closure into the the reader something like:

```rust
let reader = Reader::from("#foobar").with_transformer(|token| token);
```

This also means that parsing needs to be delayed until the reader is used as an iterator or converted to a vec.

Thanks for taking a look, let me know if there's something I can do to help.

File: lib/posticle/src/tokens.rs (hunk 1 above)
Comment (measlytwerp) [reply lib/posticle/src/tokens.rs]: Yeah, quite right, I will change them all.

**Comment by barzamin** [comment]
ahhhhh, okay. that makes a lot of sense. i didn't check, but does this PR wire everything up to rustodon core?

**Comment by barzamin** [comment]
oh, awesome! excited to actually wire this up for _posts_ soon.

**Comment by measlytwerp** [comment]
@barzamin I need a little help with this, if you take a look at `transforms.rs` you'll see that there's an issue where the `account_lookup` closure lifetime doesn't last as long as the boxed `transformer` closure. I'm not sure how to resolve this.

Once it's resolved, this PR is ready, and work can start on basic Markdown support!

**Comment by barzamin** [comment]
trying to figure out how to convince the borrowck that account_lookup outlives reader...

**Comment by barzamin** [comment]
sorry i let this stagnate :p

i'm gonna take a look at this again and see if i can wrangle the borrowck into doing what we want. at the worst, we can just `Box<>` some stuff and hope

**Comment by measlytwerp** [comment]
All things considered that really isn't hard to follow, thanks for showing me.

**Comment by barzamin** [comment]
Yeah, this looks awesome! I'm gonna merge now, :crossed_fingers:
